File: scripts/rust/benches/review_duplicate_bench_names.py
#!/usr/bin/env python3
"""
Review: No duplicate benchmark names in Cargo.toml.

Each [[bench]] entry must have a unique name.
"""
# Git commit: 584a672b6a34782766863c5f76a461d3297a741a
# Date: 2025-10-17 05:17:36 -0700


import re
import sys
from pathlib import Path
from collections import defaultdict


def main():
    repo_root = Path(__file__).parent.parent.parent.parent
    cargo_toml = repo_root / "Cargo.toml"
    
    with open(cargo_toml, 'r', encoding='utf-8') as f:
        content = f.read()
    
    # Find all [[bench]] entries with name and path
    # Pattern: [[bench]]\nname = "X"\npath = "Y"
    bench_pattern = re.compile(
        r'\[\[bench\]\]\s*name\s*=\s*"([^"]+)"\s*path\s*=\s*"([^"]+)"',
        re.MULTILINE
    )
    
    bench_entries = bench_pattern.findall(content)
    
    # Group by name to find duplicates
    by_name = defaultdict(list)
    for name, path in bench_entries:
        by_name[name].append(path)
    
    # Find duplicates
    duplicates = {name: paths for name, paths in by_name.items() if len(paths) > 1}
    
    if duplicates:
        print("✗ Found duplicate benchmark names in Cargo.toml:\n")
        for name, paths in sorted(duplicates.items()):
            print(f"  name = \"{name}\" appears {len(paths)} times:")
            for path in paths:
                print(f"    - {path}")
            print()
        
        total_violations = sum(len(paths) - 1 for paths in duplicates.values())
        print(f"Total violations: {total_violations}")
        print("\nFix: Each benchmark must have a unique name.")
        print("Suggestion: Add chapter suffix like 'BenchFooChap37' and 'BenchFooChap38'")
        return 1
    
    print(f"✓ All {len(bench_entries)} benchmark names are unique in Cargo.toml")
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: src/bin/review_duplicate_bench_names.rs
// Copyright (C) Brian G. Milnes 2025

//! Review: No duplicate benchmark names in Cargo.toml
//! 
//! Replaces: scripts/rust/benches/review_duplicate_bench_names.py
//! Rule: Each [[bench]] entry must have a unique name
//! Git commit: 584a672b6a34782766863c5f76a461d3297a741a
//! Binary: rusticate-review-duplicate-bench-names

use anyhow::Result;
use std::collections::HashMap;
use std::fs;
use std::time::Instant;


macro_rules! log {
    ($($arg:tt)*) => {{
        use std::io::Write;
        let msg = format!($($arg)*);
        println!("{}", msg);
        if let Ok(mut file) = std::fs::OpenOptions::new()
            .create(true)
            .append(true)
            .open("analyses/review_duplicate_bench_names.log")
        {
            let _ = writeln!(file, "{}", msg);
        }
    }};
}
fn main() -> Result<()> {
    let start = Instant::now();
    
    // Read Cargo.toml
    let cargo_content = fs::read_to_string("Cargo.toml")?;
    
    // Find all [[bench]] entries
    let mut bench_names: HashMap<String, Vec<String>> = HashMap::new();
    let lines: Vec<&str> = cargo_content.lines().collect();
    
    let mut i = 0;
    while i < lines.len() {
        if lines[i].trim() == "[[bench]]" {
            // Look for name and path in next few lines
            let mut name = None;
            let mut path = None;
            
            for j in (i+1)..std::cmp::min(i+10, lines.len()) {
                let line = lines[j].trim();
                
                if line.starts_with("name") {
                    if let Some(value) = line.split('=').nth(1) {
                        name = Some(value.trim().trim_matches('"').to_string());
                    }
                } else if line.starts_with("path") {
                    if let Some(value) = line.split('=').nth(1) {
                        path = Some(value.trim().trim_matches('"').to_string());
                    }
                }
                
                // Stop at next section
                if line.starts_with('[') && line != "[[bench]]" {
                    break;
                }
            }
            
            if let (Some(n), Some(p)) = (name, path) {
                bench_names.entry(n).or_default().push(p);
            }
        }
        i += 1;
    }
    
    // Count total benches first
    let total_benches: usize = bench_names.values().map(|v| v.len()).sum();
    
    // Find duplicates
    let mut duplicates: Vec<(String, Vec<String>)> = bench_names
        .into_iter()
        .filter(|(_, paths)| paths.len() > 1)
        .collect();
    
    duplicates.sort_by(|a, b| a.0.cmp(&b.0));
    
    if !duplicates.is_empty() {
        log!("✗ Found duplicate benchmark names in Cargo.toml:");
        log!("");
        
        let mut total_violations = 0;
        for (name, paths) in &duplicates {
            log!("  name = \"{}\" appears {} times:", name, paths.len());
            for path in paths {
                log!("    - {}", path);
            }
            log!("");
            total_violations += paths.len() - 1;
        }
        
        log!("Total violations: {}", total_violations);
        log!("\nFix: Each benchmark must have a unique name.");
        log!("Suggestion: Add chapter suffix like 'BenchFooChap37' and 'BenchFooChap38'");
        
        let elapsed = start.elapsed().as_millis();
        log!("Completed in {}ms", elapsed);
        
        std::process::exit(1);
    }
    log!("✓ All {} benchmark names are unique in Cargo.toml", total_benches);
    
    let elapsed = start.elapsed().as_millis();
    log!("Completed in {}ms", elapsed);
    
    Ok(())
}

